core/drivers/crypto/caam/caam_jr.c

```diff
@@ -36,7 +36,7 @@ struct inring_entry {
 struct __packed outring_entry {
 	uint64_t desc;   /* Physical address of the descriptor */
 	uint32_t status; /* Status of the executed job */
-};
+} __aligned(__alignof__(void *));
```

Comment:
With only `__packed` I supposed that there would be no padding and that the structs if put in an array would be without padding between each other.
But what is it now?

Reply:
Arithmetic on `void *` and hence the treatment of `void *` to be the same as `char *` is also a GNU C extension, according to [godbolt](https://godbolt.org/z/X77__p) its however also supported on clang.

Reply:
I expect that alignment of `void *` is `4` or `8` depending on AArch32 vs AArch64.

Reply:
That is indeed the case. It still looked weird to me so I double checked.

Reply:
@jenswi-linaro @Emantor 
You guys are right. After your comments we did some testing, and the ```__aligned``` attribute introduces a padding between each elements of the array on aarch64. This not wanted obviously.

A correct solution we found to make the compiler happy about the ```packed``` structure is to change input variable type of functions ```caam_desc_pop()``` and ```caam_read_jobstatus()``` like so:
```uint32_t caam_read_jobstatus(void *addr)```
```paddr_t caam_desc_pop(void *out_entry)```

What do you think?

Reply:
Would it make sense to instead supply a pointer to the `struct outring_entry`? That way the attributes of the struct can be taken into account by the function when extracting the field.
Replacing it with a void pointer just hides the warning without addressing the potential problem.
You might also consider splitting the `desc` field into two 32-bit fields since you'll otherwise have unaligned 64-bit values and this may become a problem on 64-bit platforms.

Reply:
I forgot to mention that we would off course remove the ```__aligned``` attribute. The array would remain packed like it was before, so we should not have any issue on aarch64 with ```void *``` pointer.

We could indeed supply a pointer to ```struct outring_entry```. To make things clean, we should extend this idea to ```struct inring_entry``` also.

Reply:
That's the point with this warning if I've understood it correctly. When to take the pointer of something inside a packed struct it may not be properly aligned for a generic pointer access. If accessed just as a field in a struct the compiler know how to load the field in a safe way.

Have you tried running this on a 64-bit platform?

> We could indeed supply a pointer to struct outring_entry. To make things clean, we should extend this idea to struct inring_entry also.

Yeah, I think this is the way forward.

Reply:
> 
> 
> That's the point with this warning if I've understood it correctly. When to take the pointer of something inside a packed struct it may not be properly aligned for a generic pointer access. If accessed just as a field in a struct the compiler know how to load the field in a safe way.
> 
> Have you tried running this on a 64-bit platform?

Yes, using a ```void *``` and only a ```__packed``` attribute works as expected on 32 and 64 bits platforms.

> 
> > We could indeed supply a pointer to struct outring_entry. To make things clean, we should extend this idea to struct inring_entry also.
> 
> Yeah, I think this is the way forward.

I agree with you. I will close this PR and create another one with a proper fix.

Reply:
> Yes, using a void * and only a __packed attribute works as expected on 32 and 64 bits platforms.

We must be talking about different things, because `ldr x0, <4 byte aligned address>` will cause an alignment fault.